iOS_SDK/OneSignalSDK/Source/OneSignal.m

```diff
@@ -2523,54 +2529,110 @@ + (id)getTriggerValueForKey:(NSString *)key {
-
+    
     // return if the user has not granted privacy permissions
     if ([self shouldLogMissingPrivacyConsentErrorWithMethodName:@"setExternalUserId:"])
         return;
+    
+    [self setExternalUserId:externalId withCompletion:nil];
+}
 
-    if ([self.existingExternalUserId isEqualToString:externalId])
++ (void)setExternalUserId:(NSString * _Nonnull)externalId withCompletion:(OSUpdateExternalUserIdBlock _Nullable)completionBlock {
+    
+    // return if the user has not granted privacy permissions
+    if ([self shouldLogMissingPrivacyConsentErrorWithMethodName:@"setExternalUserId:withCompletion:"])
         return;
 
+    // Can't set the external id if init is not done or the app id or user id has not ben set yet
     if (!performedOnSessionRequest) {
         // will be sent as part of the registration/on_session request
         pendingExternalUserId = externalId;
         return;
     } else if (!self.currentSubscriptionState.userId || !self.app_id) {
-        [self onesignal_Log:ONE_S_LL_WARN message:[NSString stringWithFormat:@"Attempted to set external-userID while %@ is not set", self.app_id == nil ? @"app ID" : @"OneSignal user ID"]];
+        [OneSignal onesignal_Log:ONE_S_LL_WARN message:[NSString stringWithFormat:@"Attempted to set external user id, but %@ is not set", self.app_id == nil ? @"app_id" : @"user_id"]];
         return;
     }
-
+    
+    // Begin constructing the request for the external id update
     let requests = [NSMutableDictionary new];
     requests[@"push"] = [OSRequestUpdateExternalUserId withUserId:externalId withOneSignalUserId:self.currentSubscriptionState.userId appId:self.app_id];
-
-    if (self.currentEmailSubscriptionState.emailUserId && (self.currentEmailSubscriptionState.requiresEmailAuth == false || self.currentEmailSubscriptionState.emailAuthCode))
+    
+    // Check if the email has been set, this will decide on updtaing the external id for the email channel
+    if ([self isEmailSetup])
         requests[@"email"] = [OSRequestUpdateExternalUserId withUserId:externalId withOneSignalUserId:self.currentEmailSubscriptionState.emailUserId appId:self.app_id];
-
-    [OneSignalClient.sharedClient executeSimultaneousRequests:requests withSuccess:^(NSDictionary<NSString *,NSDictionary *> *results) {
-        // the success/fail callbacks always execute on the main thread
-        [self successfullySentExternalUserId:externalId];
-    } onFailure:^(NSDictionary<NSString *,NSError *> *errors) {
-        // if either request fails, this block is executed
-        NSError *error = errors[@"push"] ?: errors[@"email"];
-        [self onesignal_Log:ONE_S_LL_ERROR message:[NSString stringWithFormat:@"Encountered an error while attempting to set external ID: %@", error.description]];
+    
+    
+    // Make sure this is not a duplicate request, if the email and push channels are aligned correctly with the same external id
+    NSMutableDictionary *results = [NSMutableDictionary new];
```

Comment:
this seems like it can be on a different method, for readability, what do you think?

Reply:
all the if (![self shouldUpdateExternalUserId:externalId withRequests:requests])  block